### PR TITLE
Energenie sockets preserve state

### PR DESCRIPTION
**Pull request by NotBobTheBuilder**
At present when you initialise the `Energenie` class, sockets are either switched on or off according to `initial_value`.

Sometimes it is useful to initialise the class without setting a state, for example starting a service to control a socket which respects its initial state.

This PR allows specifying `None` as an `initial_value`, in which case no message is sent to the socket. 

The default behaviour of switching off the socket remains unchanged

### CODE REVIEW

**Comment by bennuttall** [comment]
Thanks Jack. I seem to remember @waveform80 intentionally not allowing this, because you can't read the initial state, so if you tried to read the value after init, before explicitly turning it on or off, it might "lie" to you.

A compromise may be to allow `initial_value` to be `None`, but then let `value` be `None` until it's been set by turning the device on or off.

I'll try and get a decision from Dave.

**Comment by bennuttall** [comment]
Sounds good to me. Push another commit and I'll try it out.

Will defer to Dave to make the final call, since we wrote it in the first place. But this seems a reasonable way of dealing with it.

**Comment by bennuttall** [comment]
Maybe look at adding tests for this edge case if you can.

**Comment by lurch** [comment]
It's actually the "other side of the coin" but you may find the discussions in #707 and #323 of interest too :wink: 

I wondered if having `value == None` might be a weird thing for the GpioZero API to support, but then I remembered that we use it in servo https://gpiozero.readthedocs.io/en/stable/api_output.html#gpiozero.Servo.value
"This can also be the special value None indicating that the servo is currently “uncontrolled”, i.e. that no control signal is being sent."

So I guess it might make sense that for Energenie "This can also be the special value None indicating that the state of the socket is unknown"? And obviously while you could read a `None` value from an Energenie, you should be prevented from setting `.value` to None.

File: gpiozero/boards.py [hunk 1]
```diff
@@ -2045,11 +2045,13 @@ class Energenie(SourceMixin, Device):
...
 
-    :param bool initial_value:
+    :type initial_value: bool or None
+    :param initial_value:
         The initial state of the socket. As Energenie sockets provide no
-        means of reading their state, you must provide an initial state for
+        means of reading their state, you may provide an initial state for
         the socket, which will be set upon construction. This defaults to
         :data:`False` which will switch the socket off.
+        Specifying :data:`None` will not switch the socket on construction.
```
Comment (lurch) [inline gpiozero/boards.py:2054]: "will not alter the socket's state" might be less ambiguous than "will not switch the socket on construction" ?

("not switch the socket on" could be misinterpreted as "switch the socket off" ? :wink: )

File: gpiozero/boards.py [hunk 2]
```diff
@@ -2063,15 +2065,15 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
...
+        if initial_value not in (True, False, None):
+            raise EnergenieBadInitialValue("initial value must be bool or None")
         self._value = None
         super(Energenie, self).__init__(pin_factory=pin_factory)
         self._socket = socket
         self._master = _EnergenieMaster(pin_factory=pin_factory)
         if initial_value:
             self.on()
-        else:
+        elif initial_value is False:
```
Comment (lurch) [inline gpiozero/boards.py:2076]: I don't think we should be checking explicitly for True or False, but instead simply checking for None, and then just using duck-typing otherwise? See e.g. https://github.com/gpiozero/gpiozero/blob/master/gpiozero/input_devices.py#L112

I think this also means the `if initial_value not in (True, False, None):` line would need to be removed?

File: gpiozero/boards.py [hunk 3]
```diff
@@ -2102,6 +2105,9 @@ def value(self):
         """
         Returns :data:`True` if the socket is on and :data:`False` if the
         socket is off.  Setting this property changes the state of the socket.
+        Returns :data:`None` only when constructedwith :data:`initial_value`
```
Comment (lurch) [inline gpiozero/boards.py:2108]: Missing space in "constructedwith" :wink:

File: gpiozero/boards.py [hunk 4]
```diff
@@ -2102,6 +2105,9 @@ def value(self):
...
+        set to :data:`None` and neither :data:`on()` nor :data:`off()` have
+        been called since construction
```
Comment (lurch) [inline gpiozero/boards.py:2110]: Missing full-stop at end of sentence.

File: gpiozero/boards.py [hunk 5]
```diff
@@ -2063,13 +2066,13 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
...
             raise EnergenieBadSocket('socket number must be between 1 and 4')
-        if initial_value is None:
-            raise EnergenieBadInitialValue("initial value can't be None")
         self._value = None
         super(Energenie, self).__init__(pin_factory=pin_factory)
         self._socket = socket
         self._master = _EnergenieMaster(pin_factory=pin_factory)
-        if initial_value:
+        if initial_value is None:
+            pass
```
Comment (lurch) [inline gpiozero/boards.py:2074]: Ugh :slightly_frowning_face: 
Could you please make this:
```python3
if initial_value is not None:
    if initial_value:
        self.on()
    else:
        self.off()
```

File: tests/test_boards.py [hunk 6]
```diff
@@ -1295,6 +1297,15 @@ def test_energenie(mock_factory):
...
         pins[5].assert_states_and_times([(0.0, False), (0.1, True), (0.25, False)])
+        [pin.clear_states() for pin in pins]
+        device3.on()
+        assert device3.value
+        pins[0].assert_states_and_times([(0.0, False), (0.0, True)])
+        pins[1].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[2].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[3].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[4].assert_states_and_times([(0.0, False)])
+        pins[5].assert_states_and_times([(0.0, False), (0.1, True), (0.25, False)])
```
Comment (lurch) [inline tests/test_boards.py:1308]: Would you mind adding tests that `device3.value = 0` and `device3.value = 1` work as expected, and `device3.value = None` raises an exception? Thanks!

File: gpiozero/boards.py [hunk 7]
```diff
@@ -2063,13 +2066,13 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
             raise EnergenieSocketMissing('socket number must be provided')
         if not (1 <= socket <= 4):
             raise EnergenieBadSocket('socket number must be between 1 and 4')
-        if initial_value is None:
-            raise EnergenieBadInitialValue("initial value can't be None")
```
Comment (lurch) [inline gpiozero/boards.py:2067]: Just for the sake of tidiness, perhaps it's worth removing this exception from the library? :shrug:

Comment (lurch) [reply gpiozero/boards.py:2067]: ...and also from `exc.py` ? (sorry for not being more explicit)

Comment (NotBobTheBuilder) [reply gpiozero/boards.py:2067]: Should be removed :) (maybe you are seeing a partial diff?)

File: gpiozero/boards.py [hunk 8]
```diff
@@ -2102,11 +2102,16 @@ def value(self):
...
+        Returns :data:`None` only when constructed with :data:`initial_value`
+        set to :data:`None` and neither :data:`on()` nor :data:`off()` have
+        been called since construction.
         """
         return self._value
 
     @value.setter
     def value(self, value):
+        if value is None:
+            raise TypeError('value cannot be None')
```
Comment (lurch) [inline gpiozero/boards.py:2114]: A more specific exception would be better - hmm perhaps `EnergenieBadInitialValue` in `exc.py` could be renamed to `EnergenieBadValue` (with the detail-string updated too) and then used here? :slightly_smiling_face:

File: tests/test_boards.py [hunk 9]
```diff
@@ -1295,6 +1297,25 @@ def test_energenie(mock_factory):
...
+
+        device3.value = False
+        device3.value = True
+        try:
+            device3.value = None
+        except:
```
Comment (lurch) [inline tests/test_boards.py:1314]: Should probably use the `with pytest.raises(ValueError)` syntax rather than using try / except?

File: tests/test_boards.py [hunk 10]
```diff
@@ -1295,6 +1297,25 @@ def test_energenie(mock_factory):
...
+        assert device3.value
+        pins[0].assert_states_and_times([(0.0, False), (0.0, True)])
+        pins[1].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[2].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[3].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[4].assert_states_and_times([(0.0, False)])
+        pins[5].assert_states_and_times([(0.0, False), (0.1, True), (0.25, False)])
+
+        device3.value = False
+        device3.value = True
```
Comment (lurch) [inline tests/test_boards.py:1311]: I'd probably make this:
```python3
        device3.value = False
        assert not device3.value
        device3.value = True
        assert device3.value
```
(i.e. check that the value you set is also the value you read back!)

File: gpiozero/boards.py [hunk 11]
```diff
@@ -2063,15 +2065,13 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
...
-        if initial_value is None:
-            raise EnergenieBadInitialValue("initial value can't be None")
         self._value = None
         super(Energenie, self).__init__(pin_factory=pin_factory)
         self._socket = socket
         self._master = _EnergenieMaster(pin_factory=pin_factory)
         if initial_value:
             self.on()
-        else:
+        elif initial_value is not None:
```
Comment (lurch) [inline gpiozero/boards.py:2074]: Sorry to be such a nitpicker, but although what you're proposing here is "more compact", I still prefer the "more explicit" version that I suggested:
```python3
if initial_value is not None:
    if initial_value:
        self.on()
    else:
        self.off()
```

I guess it's up to @bennuttall really though!

**Comment by bennuttall** [comment]
Dave approves!
